### DISCUSSION TITLE
[Feature] NOT operator

### DISCUSSION THREAD

**Discussion by Alihkhawaher** [immich-app/immich#10381] [Feature Request]
### I have searched the existing feature requests to make sure this is not a duplicate request.

- [X] Yes

### The feature

I want to use  "NOT" or "!" when searching. Examples, I want to search for a photo that is "NOT person" or NOT Samsung. 
Many thanks,

### Platform

- [X] Server
- [X] Web
- [X] Mobile

**Comment by qkrrmsdud1** [comment]
Seconding this. Really useful for purging unwanted assets.

**Comment by hwaterke** [comment]
I recently opened a PR going in that direction https://github.com/immich-app/immich/pull/18552 (starting with albums)
(it only targeted the server and not yet the frontends)

@mertalev suggested that more thought should be put into the structure of the API with regards to exclusion patterns.
I've just opened a focus discussion on Discord on this topic: https://discord.com/channels/979116623879368755/1376602421064634460

**Comment by photoserver-pro** [comment]
Thanks for raising this feature request — I strongly agree with the direction you’re proposing.

For larger libraries especially, being able to refine results directly within the current view (instead of restarting or rebuilding searches) would significantly improve usability and discoverability. Immich’s search performance is already excellent; adding more flexible filtering and refinement on top of that would unlock a lot of additional value without changing the core concepts.
In particular, the existing and potential use of search operators feels like a very strong foundation for a future faceted / filter-based navigation. 
From a user perspective, this kind of iterative narrowing is very intuitive and well understood from other domains (e.g. large catalogs or media libraries).

**Comment by Romiko** [comment]
+1 - Happy to help with the coding - can use something like [lucene indexing ](https://lucene.apache.org/core/)

**Comment by sbiden** [comment]
As a user, I just want to have filters similar to Boolean and logical operations, and if translated into a customer UX

1) find all photos with tags: a b c d- ALL photos should not contain all these tags.

2) find all photos that contain all these tags: a b c d
)
3) Find all photos with the tag "Album_tag_A" that contain everything and without ( like NOT operator) "tag_car"

**Comment by photoserver-pro** [comment]
@sbiden These examples show exactly why filtering needs to be treated as a first-class navigation primitive — not just as a search refinement tool, but as the primary way to make large result sets explorable. Boolean operators like NOT are a natural extension of that: the ability to iteratively narrow a result set without restarting the search from scratch.

I've written a broader piece on how filtering, sorting, and temporal navigation need to work together as a consistent set of primitives across all views in Immich — search results, people, tags, locations, albums, and folders: https://github.com/immich-app/immich/discussions/27151

Would love to hear how others in this thread see the NOT operator fitting into a wider filtering model.

**Comment by fernplant** [comment]
Agree. I was hoping to do a search for photos containing X pereson but not Y person. Doesnt seem possible at the moment. Boolean operators for search please

**Comment by Luke-Shepp** [comment]
+1

For example, "all images in dog album that do not contain a dog" - so that I can clean up my albums!

**Comment by timonrieger** [comment]
#30179 solves this. look into the examples there to see what's gonna be possible
